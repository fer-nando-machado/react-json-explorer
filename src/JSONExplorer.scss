.json-explorer {
  font-family: Menlo, Monaco, Consolas, "Courier New", monospace;
  overflow-x: auto;
  display: flex;
  flex-direction: column;

  input {
    font-family: Menlo, Monaco, Consolas, "Courier New", monospace;
    font-size: 16px;
    padding: 10px;
    margin: 0 4px;
    border-radius: 12px;
    outline: 0;
    height: 20px;
    border: 2px solid initial;

    &:focus {
      border: 2px solid blue;
    }
  }

  aside {
    color: gray;
    margin: 0 12px;
    height: 14px;
  }

  ul,
  li {
    list-style-type: none;
    padding-left: 12px;
  }

  ul {
    &.array {
      &::before {
        content: "[";
      }
      &::after {
        content: "],";
      }
    }

    &.object {
      &::before {
        content: "{";
      }
      &::after {
        content: "},";
      }
    }
  }

  .key {
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;

    &::after {
      color: black;
      content: ":";
      font-weight: normal;
      padding-right: 6px;
    }

    &.clickable {
      color: blue;
      &:hover {
        cursor: pointer;
        font-weight: bold;
        text-decoration: underline;
      }
    }
  }

  .value {
    white-space: nowrap;

    &::after {
      color: black;
      content: ",";
    }

    &.string {
      color: orange;
      &::before {
        color: black;
        content: "'";
      }
      &::after {
        color: black;
        content: "',";
      }
    }

    &.boolean {
      color: brown;
    }

    &.number {
      color: forestgreen;
    }

    &.undefined {
      color: gray;
    }

    &.object {
      color: hotpink;
    }
  }
}
