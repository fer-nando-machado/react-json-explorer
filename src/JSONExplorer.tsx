import React, { useEffect, useMemo, useState } from "react";
import "./JSONExplorer.scss";

type JSONValue =
  | string
  | number
  | boolean
  | JSONValue[]
  | undefined
  | null
  | { [key: string]: JSONValue };
type JSONObject = { [key: string]: JSONValue };

type JSONExplorerProps = {
  data: JSONObject;
};

const cachedPathValue = new Map<string, string>();

const renderKey = (
  key: string,
  path: string,
  onClick?: (path: string) => void
) => {
  return (
    <span
      className={`key ${onClick ? "clickable" : ""}`}
      onClick={onClick ? () => onClick(path) : undefined}
    >
      {key}
    </span>
  );
};

const buildPath = (key: string, path?: string, isArrayIndex?: boolean) => {
  if (!path) {
    return key;
  }
  if (isArrayIndex) {
    return `${path}[${key}]`;
  }
  return `${path}.${key}`;
};

const renderValue = (
  key: string,
  value: JSONValue,
  path?: string,
  onClick?: (path: string) => void,
  isArrayIndex?: boolean
) => {
  const typeOfValue = typeof value;
  const currentPath = buildPath(key, path, isArrayIndex);

  if (Array.isArray(value)) {
    return (
      <li key={key}>
        {renderKey(key, currentPath)}
        <ul className="array">
          {value.map((item, index) =>
            renderValue(index.toString(), item, currentPath, onClick, true)
          )}
        </ul>
      </li>
    );
  }

  if (typeOfValue === "object" && value !== null) {
    const objectList = (
      <ul className="object">
        {Object.entries(value as Object).map(([subKey, subValue]) =>
          renderValue(subKey, subValue, currentPath, onClick)
        )}
      </ul>
    );
    return path ? <li key={key}>{objectList}</li> : objectList;
  }

  const valueString = String(value);
  cachedPathValue.set(currentPath, valueString);
  return (
    <li key={key}>
      {renderKey(key, currentPath, onClick)}
      <span className={`value ${typeOfValue}`}>{valueString}</span>
    </li>
  );
};

const JSONExplorer: React.FC<JSONExplorerProps> = ({ data }) => {
  const [property, setProperty] = useState("");
  const [value, setValue] = useState("");

  const onClick = (path: string) => {
    setProperty(path);
  };

  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setProperty(event.target.value);
  };

  useEffect(() => {
    if (!property) {
      setValue("");
      return;
    }
    const innerText = cachedPathValue.get(property);
    setValue(innerText || "undefined");
  }, [property]);

  const dataTree = useMemo(() => {
    cachedPathValue.clear();
    return renderValue("data", data, undefined, onClick);
  }, [data]);

  return (
    <div className="json-explorer">
      <input type="text" value={property} onChange={onChange}></input>
      <label>{value}</label>
      {dataTree}
    </div>
  );
};

export default JSONExplorer;
