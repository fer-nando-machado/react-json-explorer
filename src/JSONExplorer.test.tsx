import "@testing-library/react";
import "@testing-library/jest-dom";
import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import JSONExplorer from "./JSONExplorer";

describe("JSONExplorer", () => {
  it("should render the component correctly", () => {
    const data = {
      date: "2021-10-27T07:49:14.896Z",
      hasError: false,
      token: 777,
      fruits: ["banana", "apple", "orange"],
      fields: [
        {
          id: "4c212130",
          prop: "iban",
          value: "DE81200505501265402568",
          hasError: false,
        },
        {
          amount: 7.77,
          size: undefined,
          hasError: true,
        },
      ],
    };

    const { container } = render(<JSONExplorer data={data} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
