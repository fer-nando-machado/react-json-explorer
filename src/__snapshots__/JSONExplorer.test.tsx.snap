// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JSONExplorer > should render the component correctly 1`] = `
<div
  class="json-explorer"
>
  <input
    type="text"
    value=""
  />
  <label />
  <ul
    class="object"
  >
    <li>
      <span
        class="key clickable"
      >
        date
      </span>
      <span
        class="value string"
      >
        2021-10-27T07:49:14.896Z
      </span>
    </li>
    <li>
      <span
        class="key clickable"
      >
        hasError
      </span>
      <span
        class="value boolean"
      >
        false
      </span>
    </li>
    <li>
      <span
        class="key clickable"
      >
        token
      </span>
      <span
        class="value number"
      >
        777
      </span>
    </li>
    <li>
      <span
        class="key "
      >
        fruits
      </span>
      <ul
        class="array"
      >
        <li>
          <span
            class="key clickable"
          >
            0
          </span>
          <span
            class="value string"
          >
            banana
          </span>
        </li>
        <li>
          <span
            class="key clickable"
          >
            1
          </span>
          <span
            class="value string"
          >
            apple
          </span>
        </li>
        <li>
          <span
            class="key clickable"
          >
            2
          </span>
          <span
            class="value string"
          >
            orange
          </span>
        </li>
      </ul>
    </li>
    <li>
      <span
        class="key "
      >
        fields
      </span>
      <ul
        class="array"
      >
        <li>
          <ul
            class="object"
          >
            <li>
              <span
                class="key clickable"
              >
                id
              </span>
              <span
                class="value string"
              >
                4c212130
              </span>
            </li>
            <li>
              <span
                class="key clickable"
              >
                prop
              </span>
              <span
                class="value string"
              >
                iban
              </span>
            </li>
            <li>
              <span
                class="key clickable"
              >
                value
              </span>
              <span
                class="value string"
              >
                DE81200505501265402568
              </span>
            </li>
            <li>
              <span
                class="key clickable"
              >
                hasError
              </span>
              <span
                class="value boolean"
              >
                false
              </span>
            </li>
            <li>
              <span
                class="key clickable"
              >
                ref
              </span>
              <span
                class="value object"
              >
                null
              </span>
            </li>
          </ul>
        </li>
        <li>
          <ul
            class="object"
          >
            <li>
              <span
                class="key clickable"
              >
                amount
              </span>
              <span
                class="value number"
              >
                7.77
              </span>
            </li>
            <li>
              <span
                class="key clickable"
              >
                size
              </span>
              <span
                class="value undefined"
              >
                undefined
              </span>
            </li>
            <li>
              <span
                class="key clickable"
              >
                hasError
              </span>
              <span
                class="value boolean"
              >
                true
              </span>
            </li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
